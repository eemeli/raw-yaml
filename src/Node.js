export const LOG = false

export class Range {
  constructor (start, end) {
    this.start = start
    this.end = end || start
  }
}

export default class Node {
  static Prop = {
    ANCHOR: 'ANCHOR',
    TAG: 'TAG'
  }

  constructor (src) {
    this.src = src
    this.nodeRange = null
    this.valueRange = null
    this.commentRange = null
    this.anchor = null
    this.tag = null
  }

  get comment () {
    if (!this.src || !this.commentRange) return null
    const { start, end } = this.commentRange
    return this.src.slice(start, end)
  }

  get rawValue () {
    if (!this.src || !this.valueRange) return null
    const { start, end } = this.valueRange
    return this.src.slice(start, end)
  }

  endIndent (offset) {
    let ch = this.src[offset]
    while (ch && ch === ' ') ch = this.src[offset += 1]
    return offset
  }

  endLine (offset) {
    let ch = this.src[offset]
    while (ch && ch !== '\n') ch = this.src[offset += 1]
    return offset
  }

  endWhiteSpace (offset) {
    let ch = this.src[offset]
    while (ch && (ch === '\t' || ch === ' ')) ch = this.src[offset += 1]
    return offset
  }

  parseComment (offset) {
    if (this.src[offset] === '#') {
      const start = offset + 1
      const end = this.endLine(start)
      this.commentRange = new Range(start, end)
      LOG && console.log('comment', this.commentRange)
      return end
    }
    LOG && console.log('comment skip')
    return offset
  }

  parseProps (offset) {
    let ch = this.src[offset]
    while (ch === '&' || ch === '!') {
      const end = this.endIdentifier(offset)
      const prop = ch === '&' ? 'anchor' : 'tag'
      this[prop] = this.src.slice(offset + 1, end)
      offset = this.endWhiteSpace(end)
      ch = this.src[offset]
      LOG && console.log('props', { prop, value: this[prop], offset, ch })
    }
    return offset
  }

}
