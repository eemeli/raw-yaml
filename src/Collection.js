import Node from './Node'
import Range from './Range'
import Scalar from './Scalar'

export default class Collection extends Node {
  constructor (doc, props) {
    super(doc, props)
    this.items = []
  }

  parse (start, indent, inFlow) {
    trace: ({ start, indent })
    const { src } = this.doc
    let offset = Node.endOfWhiteSpace(src, start)
    let ch = src[offset]
    let lineStart = start - indent
    let valueEnd = offset
    while (ch) {
      while (ch === '\n' || ch === '#') {
        if (ch === '#') {
          const comment = new Scalar(this.doc, { type: Node.Type.COMMENT })
          offset = comment.parse(offset, indent, true)
          this.items.push(comment)
          valueEnd = comment.commentRange.end
          if (offset >= src.length) break
        }
        lineStart = offset + 1
        offset = Node.endOfIndent(src, lineStart)
        if (Node.isBlank(src, offset)) {
          const wsEnd = Node.endOfWhiteSpace(src, offset)
          const next = src[wsEnd]
          if (!next || next === '\n' || next === '#') {
            offset = next
          }
        }
        ch = src[offset]
      }
      if (!ch || offset !== lineStart + indent) break
      trace: 'item-start', this.items.length, ch
      const node = this.doc.parseNode(offset, indent, inFlow)
      this.items.push(node)
      valueEnd = node.valueRange.end
      // FIXME: prevents infinite loop
      if (node.range.end <= offset) throw new Error(`empty node ${node.type} ${JSON.stringify(node.range)}`)
      offset = Node.endOfWhiteSpace(src, node.range.end)
      ch = src[offset]
    }
    this.valueRange = new Range(start, valueEnd)
    trace: 'items', this.items
    return offset
  }
}
