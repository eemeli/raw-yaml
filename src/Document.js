import Node from './Node'
import Range from './Range'

export default class Document extends Node {
  static endOfDirective (src, offset) {
    let ch = src[offset]
    while (ch && ch !== '\n' && ch !== '#') ch = src[offset += 1]
    // last char can't be whitespace
    ch = src[offset - 1]
    while (ch === ' ' || ch === '\t') {
      offset -= 1
      ch = src[offset - 1]
    }
    return offset
  }

  static startCommentOrEndBlankLine (src, start) {
    const offset = Node.endOfWhiteSpace(src, start)
    const ch = src[offset]
    return ch === '#' || ch === '\n' ? offset : start
  }

  static atDirectivesEnd (src, offset) {
    return src[offset] === '-' && src[offset + 1] === '-' && src[offset + 2] === '-'
  }

  static atDocumentEnd (src, offset) {
    return !src[offset] || (
      src[offset] === '.' && src[offset + 1] === '.' && src[offset + 2] === '.'
    )
  }

  constructor (props) {
    super(props)
    this.directives = null
    this.contents = null
  }

  parseDirectives (start) {
    const { src } = this.context
    this.directives = []
    let offset = start
    while (!Document.atDirectivesEnd(src, offset)) {
      offset = Document.startCommentOrEndBlankLine(src, offset)
      switch (src[offset]) {
        case '\n':
          offset += 1
          break
        case '#': {
          const comment = new Node({ type: Node.Type.COMMENT }, this.context)
          offset = comment.parseComment(offset)
          this.directives.push(comment)
          trace: 'directive-comment', comment.commentRange, JSON.stringify(comment.comment)
        } break
        case '%': {
          const dirStart = offset
          const directive = new Node({ type: Node.Type.DIRECTIVE }, this.context)
          offset = Document.endOfDirective(src, offset + 1)
          directive.valueRange = new Range(dirStart + 1, offset)
          offset = Node.endOfWhiteSpace(src, offset)
          offset = directive.parseComment(offset)
          directive.range = new Range(dirStart, offset)
          this.directives.push(directive)
          trace: 'directive', { valueRange: directive.valueRange, commentRange: directive.commentRange }, JSON.stringify(directive.rawValue)
        } break
        default:
          return offset
      }
    }
    return offset + 3
  }

  parseContents (start) {
    const { src } = this.context
    this.contents = []
    let offset = Node.endOfWhiteSpace(src, start)
    this.valueRange = new Range(offset)
    while (!Document.atDocumentEnd(src, offset)) {
      switch (src[offset]) {
        case '\n':
          offset += 1
          break
        case '#': {
          const comment = new Node({ type: Node.Type.COMMENT }, this.context)
          offset = comment.parseComment(offset)
          this.contents.push(comment)
          trace: 'content-comment', comment.commentRange, JSON.stringify(comment.comment)
        } break
        default: {
          const iEnd = Node.endOfIndent(src, offset)
          const context = { indent: iEnd - offset - 1, inFlow: false, inCollection: false, src }
          const node = this.context.parseNode(context, iEnd)
          if (!node) return offset // at next document start
          this.contents.push(node)
          this.valueRange.end = node.valueRange.end
          if (node.range.end <= offset) throw new Error(`empty node ${node.type} ${JSON.stringify(node.range)}`)
          offset = node.range.end
          trace: 'content-node', { valueRange: node.valueRange, commentRange: node.commentRange }, JSON.stringify(node.rawValue)
        }
      }
      offset = Document.startCommentOrEndBlankLine(src, offset)
    }
    return src[offset] ? offset + 3 : offset
  }

  /**
   *
   * @param {!Object} context
   * @param {!number} start - Index of first character
   * @returns {!number} - Index of the character after this
   */
  parse (context, start) {
    trace: context, { start }
    this.context = context
    const { src } = context
    let offset = this.parseDirectives(start)
    offset = this.parseContents(offset)
    return offset
  }
}
