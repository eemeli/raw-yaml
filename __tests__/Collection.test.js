import Collection from '../src/Collection'
import Document from '../src/Document'
import Node from '../src/Node'
import { cleanForSnapshot } from './common'

describe.only('simple collections', () => {
  test('seq', () => {
    const seq = `
- one value
- '2' #c1
  #c2
- !tag "three"
-\t>
 four
   five
- [{"six":7}]
#c3`
    const doc = new Document(seq)
    const node = new Collection(doc, { type: Node.Type.COLLECTION })
    const end = node.parse(0, 0, false)
    expect(node.rawValue).toBe(seq)
    expect(end).toBe(seq.length)
    expect(node.items.length).toBe(7)
    expect(cleanForSnapshot(node)).toMatchSnapshot()
  })
})
