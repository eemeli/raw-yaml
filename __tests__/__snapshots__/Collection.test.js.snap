// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple collections seq 1`] = `
Collection {
  "anchor": null,
  "commentRange": null,
  "items": Array [
    CollectionItem {
      "anchor": null,
      "commentRange": null,
      "indicator": "-",
      "item": Scalar {
        "anchor": null,
        "blockStyle": null,
        "commentRange": null,
        "range": Range {
          "end": 12,
          "start": 3,
        },
        "raw": "one value",
        "tag": null,
        "type": "PLAIN",
        "valueRange": Range {
          "end": 12,
          "start": 3,
        },
      },
      "range": Range {
        "end": 12,
        "start": 1,
      },
      "tag": null,
      "type": "SEQ_ITEM",
      "valueRange": Range {
        "end": 12,
        "start": 1,
      },
    },
    CollectionItem {
      "anchor": null,
      "commentRange": null,
      "indicator": "-",
      "item": Scalar {
        "anchor": null,
        "blockStyle": null,
        "commentRange": Range {
          "end": 22,
          "start": 20,
        },
        "range": Range {
          "end": 22,
          "start": 15,
        },
        "raw": "'2'",
        "tag": null,
        "type": "SINGLE",
        "valueRange": Range {
          "end": 18,
          "start": 15,
        },
      },
      "range": Range {
        "end": 22,
        "start": 13,
      },
      "tag": null,
      "type": "SEQ_ITEM",
      "valueRange": Range {
        "end": 18,
        "start": 13,
      },
    },
    Scalar {
      "anchor": null,
      "blockStyle": null,
      "commentRange": Range {
        "end": 28,
        "start": 26,
      },
      "range": null,
      "raw": "",
      "tag": null,
      "type": "COMMENT",
      "valueRange": Range {
        "end": 25,
        "start": 25,
      },
    },
    CollectionItem {
      "anchor": null,
      "commentRange": null,
      "indicator": "-",
      "item": Scalar {
        "anchor": null,
        "blockStyle": null,
        "commentRange": null,
        "range": Range {
          "end": 43,
          "start": 31,
        },
        "raw": "\\"three\\"",
        "tag": "tag",
        "type": "DOUBLE",
        "valueRange": Range {
          "end": 43,
          "start": 36,
        },
      },
      "range": Range {
        "end": 43,
        "start": 29,
      },
      "tag": null,
      "type": "SEQ_ITEM",
      "valueRange": Range {
        "end": 43,
        "start": 29,
      },
    },
    CollectionItem {
      "anchor": null,
      "commentRange": null,
      "indicator": "-",
      "item": Scalar {
        "anchor": null,
        "blockStyle": ">",
        "commentRange": null,
        "range": Range {
          "end": 61,
          "start": 46,
        },
        "raw": " four
   five",
        "tag": null,
        "type": "BLOCK",
        "valueRange": Range {
          "end": 61,
          "start": 48,
        },
      },
      "range": Range {
        "end": 61,
        "start": 44,
      },
      "tag": null,
      "type": "SEQ_ITEM",
      "valueRange": Range {
        "end": 61,
        "start": 44,
      },
    },
    CollectionItem {
      "anchor": null,
      "commentRange": null,
      "indicator": "-",
      "item": FlowContainer {
        "anchor": null,
        "commentRange": null,
        "items": Array [
          "[",
          FlowContainer {
            "anchor": null,
            "commentRange": null,
            "items": Array [
              "{",
              Scalar {
                "anchor": null,
                "blockStyle": null,
                "commentRange": null,
                "range": Range {
                  "end": 71,
                  "start": 66,
                },
                "raw": "\\"six\\"",
                "tag": null,
                "type": "DOUBLE",
                "valueRange": Range {
                  "end": 71,
                  "start": 66,
                },
              },
              ":",
              Scalar {
                "anchor": null,
                "blockStyle": null,
                "commentRange": null,
                "range": Range {
                  "end": 73,
                  "start": 72,
                },
                "raw": "7",
                "tag": null,
                "type": "PLAIN",
                "valueRange": Range {
                  "end": 73,
                  "start": 72,
                },
              },
              "}",
            ],
            "range": Range {
              "end": 74,
              "start": 65,
            },
            "tag": null,
            "type": "FLOW_MAP",
            "valueRange": Range {
              "end": 74,
              "start": 65,
            },
          },
          "]",
        ],
        "range": Range {
          "end": 75,
          "start": 64,
        },
        "tag": null,
        "type": "FLOW_SEQ",
        "valueRange": Range {
          "end": 75,
          "start": 64,
        },
      },
      "range": Range {
        "end": 75,
        "start": 62,
      },
      "tag": null,
      "type": "SEQ_ITEM",
      "valueRange": Range {
        "end": 75,
        "start": 62,
      },
    },
    Scalar {
      "anchor": null,
      "blockStyle": null,
      "commentRange": Range {
        "end": 79,
        "start": 77,
      },
      "range": null,
      "raw": "",
      "tag": null,
      "type": "COMMENT",
      "valueRange": Range {
        "end": 76,
        "start": 76,
      },
    },
  ],
  "range": null,
  "tag": null,
  "type": "COLLECTION",
  "valueRange": Range {
    "end": 79,
    "start": 0,
  },
}
`;
